File: Niveau 4/4 – Récursivité avancée/2 - Moins de noms.py
def backtrack(lettres, size, curr, usedLetter):
   if (len(curr) == size):
      print(curr)
      return
   
   for i in range(len(lettres)):
      if(usedLetter[i]):
         continue
      curr += lettres[i]
      usedLetter[i] = True
      backtrack(lettres, size, curr, usedLetter)
      curr = curr[:-1]
      usedLetter[i] = False
      
def main():
   nbLettre = int(input())
   lettres = input()
   size = int(input())
   
   combinaison = 1
   for i in range(size):
      combinaison *= nbLettre - i
   
   print(combinaison)
   
   curr = ""
   usedLetter = [False] * nbLettre
   
   backtrack(lettres, size, curr, usedLetter)
   
main()

File: Niveau 4/4 – Récursivité avancée/2 - Moins de noms.cpp
#include <iostream>
#include <vector>

using namespace std;

void backtrack(vector<char> &letters, int size, string &curr, vector<bool> usedLetter)
{
    if (curr.size() == size)
    {
        cout << curr << '\n';
        return;
    }

    for (int i = 0; i < letters.size(); i++)
    {
        if (usedLetter[i])
        {
            continue;
        }
        curr.push_back(letters[i]);
        usedLetter[i] = true;
        backtrack(letters, size, curr, usedLetter);
        curr.pop_back();
        usedLetter[i] = false;
    }
}

int main()
{
    int nbLettre;
    cin >> nbLettre;

    vector<char> arr(nbLettre, ' ');

    for (int i = 0; i < nbLettre; i++)
    {
        cin >> arr[i];
    }

    int size;
    cin >> size;

    // Calcul combinaison
    int combinaison = 1;
    for (int i = 0; i < size; i++)
    {
        combinaison *= nbLettre - i;
    }

    cout << combinaison << '\n';

    string curr;
    vector<bool> usedLetter(nbLettre, false);

    backtrack(arr, size, curr, usedLetter);
}